influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +115,231 @@ struct CacheValue {
+    async fn set_success(&self, path: &Path, value: CacheValue) -> Result<(), anyhow::Error> {
+        match self.map.write().await.entry(path.clone()) {
+            Entry::Occupied(mut o) => {
+                let entry = o.get_mut();
+                let tx = match &entry.state {
+                    CacheEntryState::Fetching(tx) => tx.clone(),
+                    _ => bail!("attempted to set success state on a non-fetching cache entry"),
+                };
+                let cache_value = Arc::new(value);
+                entry.state = CacheEntryState::Success(Arc::clone(&cache_value));
+                let watch_count = tx.receiver_count().min((i8::MAX - 1) as usize);
+                entry.hits.store((watch_count + 1) as i32, Ordering::SeqCst);
+                // TODO(trevor): what if size is greater than cache capacity?
+                let additional = entry.size();
+                self.used.fetch_add(additional, Ordering::SeqCst);
+                // notify any watching requests that tried to get this entry while it was
+                // fetching:
+                let _ = tx.send(Some(cache_value));
+                Ok(())
+            }
+            Entry::Vacant(_) => bail!("attempted to set success state on an empty cache entry"),
+        }
+    }
+
+    /// Remove an entry from the cache, as well as its associated size from the used capacity
+    ///
+    /// This operation has *O(N)* time complexity since the underlying index map needs to have
+    /// its elements removed with their order preserved.
+    async fn remove(&self, path: &Path) {
+        let Some(entry) = self.map.write().await.shift_remove(path) else {
+            return;
+        };
+        self.used.fetch_sub(entry.state.size(), Ordering::SeqCst);
+    }
 
-impl WeightScale<Path, CacheEntry> for CacheEntryScale {
-    fn weight(&self, key: &Path, value: &CacheEntry) -> usize {
-        key.as_ref().len() + value.size()
+    /// A cache should be pruned if it is using all of its memory capacity
+    fn should_prune(&self) -> bool {
+        let used = self.used.load(Ordering::SeqCst);
+        used >= self.capacity
+    }
+
+    /// Prune least frequently hit entries from the cache stopping once the `max_freed_amount` has
+    /// been pruned.
+    ///
+    /// Entries that  are not pruned will have their hit count decremented, while entries with a
+    /// negative hit count will be skipped since those are entries that are in the process of being
+    /// fetched.
+    ///
+    /// This performs two passes over the cache's inner map:
+    /// 1. to check for the lowest hit count
+    /// 2. to do the actual pruning
+    ///
+    /// Since the inner map is an [`IndexMap`], the order of iteration during pruning will follow
+    /// the insertion order to the map, and therefore will prune "older" entries before "newer" ones.
+    async fn prune(&self) {
+        let mut map = self.map.write().await;
+        let lowest_hit_count = map
```

Comment:
I wonder if this could end up only pruning a single entry at a time, which might cause us to have to call prune many times in order to free up enough space. Just a concern because it requires a write lock to do things. Might be better to have it free up a bit more by default?

Also, the newest entries will have the fewest hits because they just got there, while the older entries will have more hits because they've been around longer. So this looks like it will favor pruning the absolute newest entries first, which is the opposite of what we want.

I think a better algorithm for us would be something like keeping the nanosecond time of when the entry was last accessed, rather than a hit count. Set that to "now" on insertion, and update to "now" whenever it is accessed. Then you could use a regular HashMap to store everything.

When it comes time to prune, you could decide what percentage of the cache to prune (like 10%) and then loop through to determine what nanosecond time represents that time. You can figure this out efficiently with a BinaryHeap.

Once you have the value that 10% of the times are lower than, you loop through and evict all those entries.

Reply:
I had a go at this in https://github.com/influxdata/influxdb/pull/25389/commits/ada33b2b43ade3fef20c7c42a49203ed6f5fb8ec - I think time complexity for prune is worst case _O(n*log(n))_, because of the while loop calling `.pop()` on the heap (which is _O(log(n))_), but generally, that loop won't iterate _n_ times, it would iterate 0.1*n times, or whatever the memory threshold is set to.

Reply:
I was thinking of something simpler and cheaper, which was to prune 10% of the entires, not paying attention to heap size. That way you don't have to calculate based on size. You figure out how many entries 10% would be (i.e. len / 10) and then add to the heap as you iterate and pop off if one of the entries should be put in. Limiting the size of the heap keeps the cost down (except for worst case).

If you're going to sort through everything anyway (which you are if you're not limiting the size of the heap as you build), you may as well just create a vec of the PruneHeapItem and sort it and then iterate through that to get the cutoff time.

Reply:
The `prune` code has been simplified down to build a heap that is fixed at `len * prune_percent` in size, with heap items that track `hit_time` which determines whether they land in the heap or not, as well as `size` and a reference to the `path` of the item. This allows a single loop over the cache map to build the heap, followed by a single loop over the heap items to remove entires directly, vs. having to do another pass with `retain` over the map.